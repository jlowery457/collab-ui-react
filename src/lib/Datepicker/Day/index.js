import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import * as utils from '@collab-ui/react/Datepicker/date_utils';

export default class Day extends React.Component {
  static displayName = 'Day';

  handleClick = event => {
    if (!this.isDisabled() && this.props.onClick) {
      this.props.onClick(event);
    }
  };

  handleKeyPress = e => {
    const { onClick } = this.props;
    e.preventDefault();
    if (e.which === 32 || e.which === 13) {
      return onClick(e);
    } else if (e.charCode === 32 || e.charCode === 13) {
      return onClick(e);
    }
  };

  handleMouseEnter = event => {
    if (!this.isDisabled() && this.props.onMouseEnter) {
      this.props.onMouseEnter(event);
    }
  };

  isSameDay = other => utils.isSameDay(this.props.day, other);

  isDisabled = () => utils.isDayDisabled(this.props.day, this.props);

  isOutsideMonth = () => {
    return (
      this.props.month !== undefined &&
      this.props.month !== utils.getMonth(this.props.day)
    );
  };

  isKeyboardSelected = () =>
    !this.isSameDay(this.props.selected) &&
    this.isSameDay(this.props.preSelection);

  getClassNames = () => {
    return classnames(
      'cui-datepicker__day',

      {
        'cui-datepicker__day--disabled': this.isDisabled(),
        'cui-datepicker__day--selected': this.isSameDay(this.props.selected),
        'cui-datepicker__day--keyboard-selected': this.isKeyboardSelected() && !this.isSameDay(utils.newDate()),
        'cui-datepicker__day--today': this.isSameDay(utils.newDate()) && !this.props.selected,
        'cui-datepicker__day--today--not-selected': this.props.selected && this.isSameDay(utils.newDate()) && !this.isSameDay(this.props.selected),
        'cui-datepicker__day--outside-month': this.isOutsideMonth()
      }
    );
  };

  render() {
    return (
      <div className='cui-datepicker__day-container'>
        <div
          className={this.getClassNames()}
          onClick={this.handleClick}
          onKeyPress={this.handleKeyPress}
          onMouseEnter={this.handleMouseEnter}
          aria-label={`day-${utils.getDate(this.props.day)}`}
          aria-selected={this.isSameDay(this.props.selected)}
          tabIndex='0' //TODO: (jlowery) not sure what to do with this.
          role='option'
        >
          {utils.getDate(this.props.day)}
        </div>
      </div>
    );
  }
}

Day.propTypes = {
  day: PropTypes.object.isRequired,
  month: PropTypes.number,
  onClick: PropTypes.func,
  onMouseEnter: PropTypes.func,
  preSelection: PropTypes.object,
  selected: PropTypes.object,
};