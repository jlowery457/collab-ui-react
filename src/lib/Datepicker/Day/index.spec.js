import React from 'react';
import { shallow } from 'enzyme';
import Day from '../Day';
import moment from 'moment';

const TEST_DATE = '2018-04-01';

describe('tests for <Day />', () => {
  const day = moment(TEST_DATE);
  it('should match day SnapShot', () => {
    const container = shallow(
      <Day day={day} month={3} />
    );

    expect(container).toMatchSnapshot();
  });

  it('should match selected day SnapShot', () => {
    const selected = moment(TEST_DATE);
    const container = shallow(
      <Day day={day} month={3} selected={selected} />
    );

    expect(container).toMatchSnapshot();
  });

  it('should render a div', () => {
    const container = shallow(
      <Day day={day} month={3} />
    );
    expect(container.find('div.cui-datepicker__day-container')).toHaveLength(1);
    expect(container.find('div.cui-datepicker__day')).toHaveLength(1);
  });

  it('should render a div styled properly when day is selected', () => {
    const selected = moment(TEST_DATE);
    const container = shallow(
      <Day day={day} month={3} selected={selected} />
    );
    expect(container.find('div.cui-datepicker__day-container')).toHaveLength(1);
    expect(container.find('div.cui-datepicker__day')).toHaveLength(1);
    expect(container.find('div.cui-datepicker__day--selected')).toHaveLength(1);
  });

  it('should render a div styled properly when day is disabled', () => {
    const maxDate = moment('2018-03-01');
    const container = shallow(
      <Day day={day} month={3} maxDate={maxDate} />
    );
    expect(container.find('div.cui-datepicker__day-container')).toHaveLength(1);
    expect(container.find('div.cui-datepicker__day')).toHaveLength(1);
    expect(container.find('div.cui-datepicker__day--disabled')).toHaveLength(1);
  });

  it('should render a div styled properly when day is outside current month', () => {
    const container = shallow(
      <Day day={day} month={4} /> // month 4 is May
    );
    expect(container.find('div.cui-datepicker__day-container')).toHaveLength(1);
    expect(container.find('div.cui-datepicker__day')).toHaveLength(1);
    expect(container.find('div.cui-datepicker__day--outside-month')).toHaveLength(1);
  });

  it('should render a div with the proper day', () => {
    const container = shallow(
      <Day day={day} month={3} /> // month 4 is May
    );
    expect(container.find('div.cui-datepicker__day').text()).toEqual('1');
  });

  it('should handle day click event', () => {
    let count = 0;
    const container = shallow(
      <Day day={day} month={3} onClick={() => count++}/>
    );

    container.find('div.cui-datepicker__day').simulate('click');
    expect(count).toEqual(1);
  });

  it('should handle day mouse enter event', () => {
    let count = 0;
    const container = shallow(
      <Day day={day} month={3} onMouseEnter={() => count++}/>
    );

    container.find('div.cui-datepicker__day').simulate('mouseenter');
    expect(count).toEqual(1);
  });
});