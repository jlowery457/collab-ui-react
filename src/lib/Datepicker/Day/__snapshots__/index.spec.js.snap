// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests for <Day /> should match day SnapShot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Day
    day={"2018-04-01T05:00:00.000Z"}
    month={3}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        aria-label="day-1"
        aria-selected={false}
        className="cui-datepicker__day"
        onClick={[Function]}
        onKeyPress={[Function]}
        onMouseEnter={[Function]}
        role="option"
        tabIndex="0"
      >
        1
      </div>,
      "className": "cui-datepicker__day-container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-label": "day-1",
        "aria-selected": false,
        "children": 1,
        "className": "cui-datepicker__day",
        "onClick": [Function],
        "onKeyPress": [Function],
        "onMouseEnter": [Function],
        "role": "option",
        "tabIndex": "0",
      },
      "ref": null,
      "rendered": 1,
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          aria-label="day-1"
          aria-selected={false}
          className="cui-datepicker__day"
          onClick={[Function]}
          onKeyPress={[Function]}
          onMouseEnter={[Function]}
          role="option"
          tabIndex="0"
        >
          1
        </div>,
        "className": "cui-datepicker__day-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-label": "day-1",
          "aria-selected": false,
          "children": 1,
          "className": "cui-datepicker__day",
          "onClick": [Function],
          "onKeyPress": [Function],
          "onMouseEnter": [Function],
          "role": "option",
          "tabIndex": "0",
        },
        "ref": null,
        "rendered": 1,
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;

exports[`tests for <Day /> should match selected day SnapShot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <Day
    day={"2018-04-01T05:00:00.000Z"}
    month={3}
    selected={"2018-04-01T05:00:00.000Z"}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <div
        aria-label="day-1"
        aria-selected={true}
        className="cui-datepicker__day cui-datepicker__day--selected"
        onClick={[Function]}
        onKeyPress={[Function]}
        onMouseEnter={[Function]}
        role="option"
        tabIndex="0"
      >
        1
      </div>,
      "className": "cui-datepicker__day-container",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "aria-label": "day-1",
        "aria-selected": true,
        "children": 1,
        "className": "cui-datepicker__day cui-datepicker__day--selected",
        "onClick": [Function],
        "onKeyPress": [Function],
        "onMouseEnter": [Function],
        "role": "option",
        "tabIndex": "0",
      },
      "ref": null,
      "rendered": 1,
      "type": "div",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <div
          aria-label="day-1"
          aria-selected={true}
          className="cui-datepicker__day cui-datepicker__day--selected"
          onClick={[Function]}
          onKeyPress={[Function]}
          onMouseEnter={[Function]}
          role="option"
          tabIndex="0"
        >
          1
        </div>,
        "className": "cui-datepicker__day-container",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "aria-label": "day-1",
          "aria-selected": true,
          "children": 1,
          "className": "cui-datepicker__day cui-datepicker__day--selected",
          "onClick": [Function],
          "onKeyPress": [Function],
          "onMouseEnter": [Function],
          "role": "option",
          "tabIndex": "0",
        },
        "ref": null,
        "rendered": 1,
        "type": "div",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactFifteenAdapter {
      "options": Object {
        "supportPrevContextArgumentOfComponentDidUpdate": true,
      },
    },
  },
}
`;
